TLM/TLM/Util/Extensions/NetSegmentExtensions.cs

```diff
@@ -9,5 +53,130 @@ public static class NetSegmentExtensions {
+        /// </summary>
+        /// <param name="segmentId">segment id</param>
+        /// <param name="netSegment">segment data</param>
+        /// <param name="startNode">reference node (optional)</param>
+        /// <param name="laneTypeFilter">lane type filter, lanes must match this filter mask</param>
+        /// <param name="vehicleTypeFilter">vehicle type filter, lanes must match this filter mask</param>
+        /// <param name="reverse">if true, lanes are ordered from right to left (relative to the
+        ///     segment's start node / the given node), otherwise from left to right</param>
+        /// <param name="sort">if false, no sorting takes place
+        ///     regardless of <paramref name="reverse"/></param>
+        /// <returns>sorted list of lanes for the given segment</returns>
+        public static IList<LanePos> GetSortedLanes(
+            this ref NetSegment netSegment,
+            bool? startNode,
+            NetInfo.LaneType? laneTypeFilter = null,
+            VehicleInfo.VehicleType? vehicleTypeFilter = null,
+            bool reverse = false,
+            bool sort = true) {
+            // TODO refactor together with getSegmentNumVehicleLanes, especially the vehicle type and lane type checks
+            NetManager netManager = Singleton<NetManager>.instance;
+            var laneList = new List<LanePos>();
+
+            bool inverted = (netSegment.m_flags & NetSegment.Flags.Invert) != NetSegment.Flags.None;
+
+            NetInfo.Direction? filterDir = null;
+            NetInfo.Direction sortDir = NetInfo.Direction.Forward;
+
+            if (startNode != null) {
+                filterDir = (bool)startNode
+                                ? NetInfo.Direction.Backward
+                                : NetInfo.Direction.Forward;
+                filterDir = inverted
+                                ? NetInfo.InvertDirection((NetInfo.Direction)filterDir)
+                                : filterDir;
+                sortDir = NetInfo.InvertDirection((NetInfo.Direction)filterDir);
+            } else if (inverted) {
+                sortDir = NetInfo.Direction.Backward;
+            }
+
+            if (reverse) {
+                sortDir = NetInfo.InvertDirection(sortDir);
+            }
+
+            NetInfo segmentInfo = netSegment.Info;
+            uint curLaneId = netSegment.m_lanes;
+            byte laneIndex = 0;
+
+            while (laneIndex < segmentInfo.m_lanes.Length && curLaneId != 0u) {
+                NetInfo.Lane laneInfo = segmentInfo.m_lanes[laneIndex];
+                if ((laneTypeFilter == null ||
+                     (laneInfo.m_laneType & laneTypeFilter) != NetInfo.LaneType.None) &&
+                    (vehicleTypeFilter == null || (laneInfo.m_vehicleType & vehicleTypeFilter) !=
+                     VehicleInfo.VehicleType.None) &&
```

Comment:
If `laneTypeFilter == null` we're going to be inspecting `None` lane types, which is something I don't think we ever want to do; TMPE is not interested in decoration lanes, is it?

So what about having a constant somewhere that's all the possible lane types we'd be interested in, and then set the default for `laneTypeFilter` argument to that? And similar for `vehicleTypeFilter` argument.

```csharp
public static IList<LanePos> GetSortedLanes(
    ....
    NetInfo.LaneType laneTypeFilter = DEFAULT_LANE_TYPES,
    VehicleInfo.VehicleType vehicleTypeFilter = DEFAULT_VEHICLE_TYPES,
    ....
```

With that in place, we can now simplify the code to:

```csharp
if (
    (laneInfo.m_laneType & laneTypeFilter) != NetInfo.LaneType.None) &&
    (laneInfo.m_vehicleType & vehicleTypeFilter) != VehicleInfo.VehicleType.None) &&
    ...
```

This way `None` lane/vehicle types never slip through (unless calling code specifically passes in `.None`). Also, if there's an applicable extension method for checking the flags... much more readable code.

And, although we can't short-circuit on `null` any more, we should easily make up for any perf impact due to the fact we're no longer spamming `laneList` with a bunch of `None` lanes.

Reply:
I suppose it depends on the use case. I don't know. I think its introducing too much complexity. maybe just put a warning in debug mode.

Reply:
Yup, a debug comment would be useful.